internal/controller/imageprefetch_controller.go

```diff
@@ -2,45 +2,508 @@ package controller
+		constants.OwnerImagePrefetchName:      imgPrefetch.Name,
+	})); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list NodeImageSets: %w", err)
+	}
+
+	status := calculateStatus(selectedNodes, nodeImageSets)
+	imgPrefetch.Status.DesiredNodes = status.desiredNodes
+	imgPrefetch.Status.ImagePulledNodes = status.availableNodes
+	imgPrefetch.Status.ImagePullingNodes = status.pullingNodes
+	imgPrefetch.Status.ImagePullFailedNodes = status.pullFailedNodes
+
+	if status.availableNodes == status.desiredNodes {
+		logger.Info("ImagePrefetch is ready", "name", imgPrefetch.Name)
+		meta.SetStatusCondition(&imgPrefetch.Status.Conditions, metav1.Condition{
+			Type:    ofenv1.ConditionReady,
+			Status:  metav1.ConditionTrue,
+			Reason:  "ImagePrefetchReady",
+			Message: "All nodes have the desired image",
+		})
+		result = ctrl.Result{}
+	}
+
+	if status.pullFailedNodes > 0 {
+		meta.SetStatusCondition(&imgPrefetch.Status.Conditions, metav1.Condition{
+			Type:    ofenv1.ConditionImagePullFailed,
+			Status:  metav1.ConditionTrue,
+			Reason:  "ImagePrefetchFailed",
+			Message: "some nodes failed to pull the image",
+		})
+	} else {
+		meta.SetStatusCondition(&imgPrefetch.Status.Conditions, metav1.Condition{
+			Type:    ofenv1.ConditionProgressing,
+			Status:  metav1.ConditionTrue,
+			Reason:  "ImagePrefetchProgressing",
+			Message: "Waiting for all nodes to pull the image",
+		})
+	}
+
+	return result, r.Status().Update(ctx, imgPrefetch)
+}
+
+type NodeImageSetStatus struct {
+	desiredNodes    int
+	availableNodes  int
+	pullingNodes    int
+	pullFailedNodes int
+}
+
+func calculateStatus(selectNodes []string, nodeImageSets *ofenv1.NodeImageSetList) NodeImageSetStatus {
+	status := NodeImageSetStatus{}
+	status.desiredNodes = len(selectNodes)
+
+	for _, nodeImageSet := range nodeImageSets.Items {
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageAvailable) {
+			status.availableNodes++
+		}
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageDownloadFailed) {
+			status.pullFailedNodes++
+		}
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageDownloadComplete) &&
+			!meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageAvailable) {
+			status.pullingNodes++
+		}
+	}
+
+	return status
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ImagePrefetchReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	nodeImageSetHandler := handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, obj client.Object) []ctrl.Request {
+			nodeImageSet := obj.(*ofenv1.NodeImageSet)
+			imagePrefetch := &ofenv1.ImagePrefetch{}
+			err := r.Get(ctx, types.NamespacedName{
+				Namespace: nodeImageSet.Labels[constants.OwnerImagePrefetchNamespace],
+				Name:      nodeImageSet.Labels[constants.OwnerImagePrefetchName],
+			}, imagePrefetch)
```

Comment:
There is no need to call `Get()` here.

NotFound error is checked in the `Reconcile()`.

Reply:
I changed it to use Get within Reconcile.
https://github.com/cybozu-go/ofen/pull/8/commits/91b593cd13efdb45e05591933590751994184444